Lib/shutil.py

```diff
@@ -81,6 +94,104 @@ def copyfileobj(fsrc, fdst, length=16*1024):
             break
         fdst.write(buf)
 
+def _zerocopy_osx(fsrc, fdst):
+    """Copy 2 regular mmap-like files by using high-performance
+    fcopyfile() syscall (OSX only).
+    """
+    try:
+        infd = fsrc.fileno()
+        outfd = fdst.fileno()
+    except Exception as err:
+        raise _GiveupOnZeroCopy(err)  # not a regular file
+
+    try:
+        posix._fcopyfile(infd, outfd)
+    except OSError as err:
+        if err.errno in {errno.EINVAL, errno.ENOTSUP}:
+            raise _GiveupOnZeroCopy(err)
+        else:
+            raise err from None
+
+def _zerocopy_win(fsrc, fdst):
+    """Copy 2 files by using high-performance CopyFileW (Windows only)."""
+    nt._win32copyfile(fsrc, fdst)
+
+def _zerocopy_sendfile(fsrc, fdst):
+    """Copy data from one regular mmap-like fd to another by using
+    high-performance sendfile() method.
+    This should work on Linux >= 2.6.33 and Solaris only.
+    """
+    global _HAS_SENDFILE
+    try:
+        infd = fsrc.fileno()
+        outfd = fdst.fileno()
+    except Exception as err:
+        raise _GiveupOnZeroCopy(err)  # not a regular file
+
+    # Hopefully the whole file will be copied in a single call.
+    # sendfile() is called in a loop 'till EOF is reached (0 return)
+    # so a bufsize smaller or bigger than the actual file size
+    # should not make any difference, also in case the file content
+    # changes while being copied.
+    try:
+        blocksize = max(os.fstat(infd).st_size, 2 ** 23)  # min 8MB
+    except Exception:
+        blocksize = 2 ** 27  # 128MB
+
+    offset = 0
+    while True:
+        try:
+            sent = os.sendfile(outfd, infd, offset, blocksize)
+        except OSError as err:
+            if err.errno == errno.ENOTSOCK:
```

Comment:
This path will also be hit consistently on FreeBSD and other systems that offer `os.sendfile`, but only with regular socket targets.

To avoid a performance regression on such platforms, it would be strongly preferred for there to be a build time configure check that told `shutil` that it shouldn't even try the optimised fast path, and should instead always go directly to `copyfileobj`.

Rather than a new flag in the OS module, this support could be indicated to the `shutil` module by exposing an `os._sendfile_copy` alias for `os.sendfile` only on platforms where `sendfile` supports regular file targets.

Reply:
How about in addition to setting the flag, we also do `_copyfileobj2 = copyfileobj`, to switch off the doomed-to-fail fastpath completely? That way after the first failed attempt, further attempts wouldn't even check the flag, since the "optimised" version would just be an alias for the regular implementation.